### PR TITLE
feat: support authentication for kstone-api

### PR DESCRIPTION
**Pull request by lianghao208**
# Test Case1: Get feature providers list without token header
- url: 127.0.0.1:8080/apis/features
- method: GET
- expect: 401 unauthorized
image

# Test Case2: login with default username and password
- url: 127.0.0.1:8080/apis/login
- method: POST
- expect: return reset password message
image

# Test Case3:  login with non-default password(after resetting password)
- url: 127.0.0.1:8080/apis/login
- method: POST
- expect: login succeeded and return JWT Token in body

login succeeded:
image

# Test Case4:  Get feature providers list with token header
- url: 127.0.0.1:8080/apis/features
- method: GET
- expect: get feature providers list succeeded

image

# Test Case5: Change user password after login
- url: 127.0.0.1:8080/apis/user
- method:PUT
- expect: change password succeeded, login with new password success

image

login success
image

# Test Case6: Add user after login
- url: 127.0.0.1:8080/apis/user
- method:POST
- expect: add user succeeded and a new user created in configmap
image

image

# Test Case7: Delete user after login
- url: 127.0.0.1:8080/apis/user
- method:DELETE
- expect: delete user succeeded and user deleted from in configmap
image
image

### CODE REVIEW

File: pkg/auth/authenticator.go [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+	Username      string `json:"username"`
+	ResetPassword bool   `json:"reset_password"`
+	Data          string `json:"data"`
+}
+
+type DefaultRequestAuthenticator struct {
+	Username   string
+	Password   string
+	PrivateKey string
+	IsLogin    bool
```
Comment (tangcong) [inline pkg/auth/authenticator.go:56]: why add IsLogin field?

File: pkg/auth/jwt.go [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+
+	"github.com/golang-jwt/jwt"
+	"k8s.io/klog/v2"
+
+	"tkestack.io/kstone/pkg/controllers/util"
+)
+
+var (
+	// defaultTTL will be used when a 'ttl' is not specified
+	defaultTTL  = 30 * time.Minute
```
Comment (tangcong) [inline pkg/auth/jwt.go:37]: too small

File: pkg/auth/authenticator.go [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+// returns a Response or an error if the request could not be checked.
+type Request interface {
+	AuthenticateRequest(ctx *gin.Context) (*Response, bool, error)
+}
+
+// Response is the struct returned by authenticator interfaces
+type Response struct {
+	Username      string `json:"username"`
+	ResetPassword bool   `json:"reset_password"`
+	Data          string `json:"data"`
```
Comment (tangcong) [inline pkg/auth/authenticator.go:49]: Username      string `json:"username"`
	ResetPassword bool   `json:"reset_password"`
	Token          string `json:"token"`
         Message string `json:"message"`

File: pkg/auth/helper.go [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+}
+
+func MiddlewareRequest(ctx *gin.Context) (*Response, bool, error) {
+	authGetter := initDefaultAuthenticatorInfoGetter(util.NewSimpleClientBuilder(""))
+	a, err := authGetter.Config(DefaultKstoneNamespace, DefaultConfigMapName, DefaultSecretName, "")
+	if err != nil {
+		klog.Errorf("get authenticator error: %v", err)
+		return unauthenticatedResponse(), false, err
+	}
+	j, err := initJWTTokenAuthenticator(
```
Comment (tangcong) [inline pkg/auth/helper.go:120]: The overall implementation needs to be more general and configurable， we will support more authenticate plugins, refer to https://github.com/tkestack/kstone/blob/master/pkg/featureprovider/plugins.go

File: pkg/auth/jwt.go (hunk 2 above)
Comment (lianghao208) [reply pkg/auth/jwt.go:37]: Any suggestions for the default value?

Comment (tangcong) [reply pkg/auth/jwt.go:37]: one day

**Pull request closed without merging**